File: ios/runanywhere-sdks-latest/sdk/runanywhere-swift/Sources/LlamaCPPRuntime/LlamaCPP.swift
//
//  LlamaCPP.swift
//  LlamaCPPRuntime Module
//
//  Unified LlamaCPP module - thin wrapper that calls C++ backend registration.
//  This replaces both LlamaCPPRuntime.swift and LlamaCPPServiceProvider.swift.
//

import CRACommons
import Foundation
import LlamaCPPBackend
import os.log
import RunAnywhere

// MARK: - LlamaCPP Module

/// LlamaCPP module for LLM text generation.
///
/// Provides large language model capabilities using llama.cpp
/// with GGUF models and Metal acceleration.
///
/// ## Registration
///
/// ```swift
/// import LlamaCPPRuntime
///
/// // Register the backend (done automatically if auto-registration is enabled)
/// try LlamaCPP.register()
/// ```
///
/// ## Usage
///
/// LLM services are accessed through the main SDK APIs - the C++ backend handles
/// service creation and lifecycle internally:
///
/// ```swift
/// // Generate text via public API
/// let response = try await RunAnywhere.chat("Hello!")
///
/// // Stream text via public API
/// for try await token in try await RunAnywhere.streamChat("Tell me a story") {
///     print(token, terminator: "")
/// }
/// ```
public enum LlamaCPP: RunAnywhereModule {
    private static let logger = SDKLogger(category: "LlamaCPP")

    // MARK: - Module Info

    /// Current version of the LlamaCPP Runtime module
    public static let version = "2.0.0"

    /// LlamaCPP library version (underlying C++ library)
    public static let llamaCppVersion = "b7199"

    // MARK: - RunAnywhereModule Conformance

    public static let moduleId = "llamacpp"
    public static let moduleName = "LlamaCPP"
    public static let capabilities: Set<SDKComponent> = [.llm]
    public static let defaultPriority: Int = 100

    /// LlamaCPP uses the llama.cpp inference framework
    public static let inferenceFramework: InferenceFramework = .llamaCpp

    // MARK: - Registration State

    private static var isRegistered = false
    private static var isVLMRegistered = false

    // MARK: - Registration

    /// Register LlamaCPP backend with the C++ service registry.
    ///
    /// This calls `rac_backend_llamacpp_register()` to register the
    /// LlamaCPP service provider with the C++ commons layer.
    /// Also registers VLM backend if available.
    ///
    /// Safe to call multiple times - subsequent calls are no-ops.
    ///
    /// - Parameter priority: Ignored (C++ uses its own priority system)
    /// - Throws: SDKError if registration fails
    @MainActor
    public static func register(priority _: Int = 100) {
        guard !isRegistered else {
            logger.debug("LlamaCPP already registered, returning")
            return
        }

        logger.info("Registering LlamaCPP backend with C++ registry...")

        // Register LLM backend
        let result = rac_backend_llamacpp_register()

        // RAC_ERROR_MODULE_ALREADY_REGISTERED is OK
        if result != RAC_SUCCESS && result != RAC_ERROR_MODULE_ALREADY_REGISTERED {
            let errorMsg = String(cString: rac_error_message(result))
            logger.error("LlamaCPP registration failed: \(errorMsg)")
            // Don't throw - registration failure shouldn't crash the app
            return
        }

        isRegistered = true
        logger.info("LlamaCPP LLM backend registered successfully")

        // Register VLM backend (Vision Language Model)
        registerVLM()
    }

    /// Register VLM (Vision Language Model) backend
    @MainActor
    private static func registerVLM() {
        guard !isVLMRegistered else { return }

        logger.info("Registering LlamaCPP VLM backend...")

        let vlmResult = rac_backend_llamacpp_vlm_register()

        if vlmResult != RAC_SUCCESS && vlmResult != RAC_ERROR_MODULE_ALREADY_REGISTERED {
            let errorMsg = String(cString: rac_error_message(vlmResult))
            logger.warning("LlamaCPP VLM registration failed: \(errorMsg) (VLM features may not be available)")
            return
        }

        isVLMRegistered = true
        logger.info("LlamaCPP VLM backend registered successfully")
    }

    /// Unregister the LlamaCPP backend from C++ registry.
    public static func unregister() {
        if isVLMRegistered {
            _ = rac_backend_llamacpp_vlm_unregister()
            isVLMRegistered = false
            logger.info("LlamaCPP VLM backend unregistered")
        }

        if isRegistered {
            _ = rac_backend_llamacpp_unregister()
            isRegistered = false
            logger.info("LlamaCPP LLM backend unregistered")
        }
    }

    // MARK: - Model Handling

    /// Check if LlamaCPP can handle a given model
    /// Uses file extension pattern matching - actual framework info is in C++ registry
    public static func canHandle(modelId: String?) -> Bool {
        guard let modelId = modelId else { return false }
        return modelId.lowercased().hasSuffix(".gguf")
    }
}

// MARK: - Auto-Registration

extension LlamaCPP {
    /// Enable auto-registration for this module.
    /// Access this property to trigger C++ backend registration.
    public static let autoRegister: Void = {
        Task { @MainActor in
            LlamaCPP.register()
        }
    }()
}

File: ios/runanywhere-sdks-latest/sdk/runanywhere-kotlin/modules/runanywhere-core-llamacpp/src/commonMain/kotlin/com/runanywhere/sdk/llm/llamacpp/LlamaCPP.kt
package com.runanywhere.sdk.llm.llamacpp

import com.runanywhere.sdk.core.module.RunAnywhereModule
import com.runanywhere.sdk.core.types.InferenceFramework
import com.runanywhere.sdk.core.types.SDKComponent
import com.runanywhere.sdk.foundation.SDKLogger

/**
 * LlamaCPP module for LLM text generation.
 *
 * Provides large language model capabilities using llama.cpp
 * with GGUF models and Metal/GPU acceleration.
 *
 * This is a thin wrapper that calls C++ backend registration.
 * All business logic is handled by the C++ commons layer.
 *
 * ## Registration
 *
 * ```kotlin
 * import com.runanywhere.sdk.llm.llamacpp.LlamaCPP
 *
 * // Register the backend (done automatically if auto-registration is enabled)
 * LlamaCPP.register()
 * ```
 *
 * ## Usage
 *
 * LLM services are accessed through the main SDK APIs - the C++ backend handles
 * service creation and lifecycle internally:
 *
 * ```kotlin
 * // Generate text via public API
 * val response = RunAnywhere.chat("Hello!")
 *
 * // Stream text via public API
 * RunAnywhere.streamChat("Tell me a story").collect { token ->
 *     print(token)
 * }
 * ```
 *
 * Matches iOS LlamaCPP.swift exactly.
 */
object LlamaCPP : RunAnywhereModule {
    private val logger = SDKLogger.llamacpp

    // MARK: - Module Info

    /** Current version of the LlamaCPP Runtime module */
    const val version = "2.0.0"

    /** LlamaCPP library version (underlying C++ library) */
    const val llamaCppVersion = "b7199"

    // MARK: - RunAnywhereModule Conformance

    override val moduleId: String = "llamacpp"

    override val moduleName: String = "LlamaCPP"

    override val capabilities: Set<SDKComponent> = setOf(SDKComponent.LLM)

    override val defaultPriority: Int = 100

    /** LlamaCPP uses the llama.cpp inference framework */
    override val inferenceFramework: InferenceFramework = InferenceFramework.LLAMA_CPP

    // MARK: - Registration State

    @Volatile
    private var isRegistered = false

    @Volatile
    private var isVlmRegistered = false

    // MARK: - Registration

    /**
     * Register LlamaCPP backend with the C++ service registry.
     *
     * This calls `rac_backend_llamacpp_register()` to register the
     * LlamaCPP service provider with the C++ commons layer, then
     * also registers the VLM (Vision Language Model) backend.
     *
     * Safe to call multiple times - subsequent calls are no-ops.
     *
     * Matches iOS LlamaCPP.register() which registers both LLM and VLM.
     *
     * @param priority Ignored (C++ uses its own priority system)
     */
    @Suppress("UNUSED_PARAMETER")
    @JvmStatic
    @JvmOverloads
    fun register(priority: Int = defaultPriority) {
        if (isRegistered) {
            logger.debug("LlamaCPP already registered, returning")
            return
        }

        logger.info("Registering LlamaCPP backend with C++ registry...")

        val result = registerNative()

        // Success or already registered is OK
        if (result != 0 && result != -4) { // RAC_ERROR_MODULE_ALREADY_REGISTERED = -4
            logger.error("LlamaCPP registration failed with code: $result")
            // Don't throw - registration failure shouldn't crash the app
            return
        }

        isRegistered = true
        logger.info("LlamaCPP LLM backend registered successfully")

        // Register VLM backend (Vision Language Model) — matches iOS pattern
        registerVLM()
    }

    /**
     * Register VLM (Vision Language Model) backend.
     * Matches iOS LlamaCPP.registerVLM() exactly.
     *
     * Wrapped in UnsatisfiedLinkError handling because the VLM JNI symbols
     * may not be present in older native library builds. In that case, LLM
     * features continue to work normally — only VLM is unavailable.
     */
    private fun registerVLM() {
        if (isVlmRegistered) return

        logger.info("Registering LlamaCPP VLM backend...")

        try {
            val vlmResult = registerVlmNative()

            if (vlmResult != 0 && vlmResult != -4) { // RAC_ERROR_MODULE_ALREADY_REGISTERED = -4
                logger.warning("LlamaCPP VLM registration failed with code: $vlmResult (VLM features may not be available)")
                return
            }

            isVlmRegistered = true
            logger.info("LlamaCPP VLM backend registered successfully")
        } catch (e: UnsatisfiedLinkError) {
            logger.warning(
                "LlamaCPP VLM native symbols not found — native library does not include VLM support. " +
                    "LLM features work normally. VLM will be available when native libs are rebuilt with VLM support.",
            )
        }
    }

    /**
     * Unregister the LlamaCPP backend from C++ registry.
     */
    fun unregister() {
        if (!isRegistered) return

        // Unregister VLM backend first (matches iOS unregister order)
        if (isVlmRegistered) {
            try {
                unregisterVlmNative()
                isVlmRegistered = false
                logger.info("LlamaCPP VLM backend unregistered")
            } catch (e: UnsatisfiedLinkError) {
                isVlmRegistered = false
                logger.warning("LlamaCPP VLM unregister skipped — native symbols not available")
            }
        }

        unregisterNative()
        isRegistered = false
        logger.info("LlamaCPP backend unregistered")
    }

    // MARK: - Model Handling

    /**
     * Check if LlamaCPP can handle a given model.
     * Uses file extension pattern matching - actual framework info is in C++ registry.
     */
    fun canHandle(modelId: String?): Boolean {
        if (modelId == null) return false
        return modelId.lowercase().endsWith(".gguf")
    }

    // MARK: - Auto-Registration

    /**
     * Enable auto-registration for this module.
     * Access this property to trigger C++ backend registration.
     */
    val autoRegister: Unit by lazy {
        register()
    }
}

/**
 * Platform-specific native registration.
 * Calls rac_backend_llamacpp_register() via JNI.
 */
internal expect fun LlamaCPP.registerNative(): Int

/**
 * Platform-specific native unregistration.
 * Calls rac_backend_llamacpp_unregister() via JNI.
 */
internal expect fun LlamaCPP.unregisterNative(): Int

/**
 * Platform-specific VLM native registration.
 * Calls rac_backend_llamacpp_vlm_register() via JNI.
 */
internal expect fun LlamaCPP.registerVlmNative(): Int

/**
 * Platform-specific VLM native unregistration.
 * Calls rac_backend_llamacpp_vlm_unregister() via JNI.
 */
internal expect fun LlamaCPP.unregisterVlmNative(): Int
